src/drivers/intel/cavs/ssp.c

```diff
@@ -126,6 +126,51 @@ static int ssp_context_restore(struct dai *dai)
 	return 0;
 }
 
+static int ssp_hw_params(struct dai *dai)
+{
+	struct ssp_pdata *ssp = dai_get_drvdata(dai);
+
+	if (!(ssp->params.clks_control &
+		SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_ES))
+		return 0;
+
+	spin_lock(&dai->lock);
+
+	/* enable port */
+	ssp_update_bits(dai, SSCR0, SSCR0_SSE, SSCR0_SSE);
+	trace_ssp("ssp_hw_params(), SSP%d port enabled", dai->index);
+
+	spin_unlock(&dai->lock);
+
+	return 0;
+}
+
+static int ssp_hw_free(struct dai *dai)
+{
+	struct ssp_pdata *ssp = dai_get_drvdata(dai);
+
+	if (!(ssp->params.clks_control &
+		SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_ES))
+		return 0;
+
+	spin_lock(&dai->lock);
+
+	trace_ssp("ssp_hw_free()");
```

Comment:
this doesn't have to be under the lock

Reply:
i think we need the lock to protect the ssp->state

Reply:
@brentlu I meant the `trace_ssp("ssp_hw_free()");` line

Reply:
it's done. thanks for review